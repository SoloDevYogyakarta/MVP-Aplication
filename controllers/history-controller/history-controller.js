"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var ServiceHistoryController_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceHistoryController = void 0;
const common_1 = require("@nestjs/common");
const lodash_1 = require("lodash");
const history_service_1 = require("../../services/history-service/history-service");
const history_repository_1 = require("../../repository/history-repository/history-repository");
const auth_guard_1 = require("../../middleware/auth-guard/auth-guard");
const platform_express_1 = require("@nestjs/platform-express");
const upload_1 = require("../../utils/upload/upload");
let ServiceHistoryController = ServiceHistoryController_1 = class ServiceHistoryController {
    constructor(repository, service) {
        this.repository = repository;
        this.service = service;
        this.logger = new common_1.Logger(ServiceHistoryController_1.name);
    }
    async create(files, req, res) {
        this.logger.log(ServiceHistoryController_1.name);
        const result = await this.service.create(req.params.id, req.body.desc, JSON.parse(req.body.data), files);
        return res.status(result.status).json((0, lodash_1.omit)(result, ['result']));
    }
    async all(res) {
        this.logger.log(ServiceHistoryController_1.name);
        const result = await this.repository.findAll();
        return res.status(common_1.HttpStatus.OK).json(result);
    }
    async detail(req, res) {
        this.logger.log(ServiceHistoryController_1.name);
        const result = await this.repository.findOne(req.params.id);
        return res.status(common_1.HttpStatus.OK).json(result);
    }
    async destroy(req, res) {
        const result = await this.service.destroy(req.params.id);
        return res.status(result.status).json(result);
    }
};
exports.ServiceHistoryController = ServiceHistoryController;
__decorate([
    (0, common_1.UseGuards)(auth_guard_1.AuthGuard),
    (0, common_1.Post)(':id'),
    (0, common_1.UseInterceptors)((0, platform_express_1.FilesInterceptor)('file', 5, { storage: upload_1.uploadOptions })),
    (0, common_1.HttpCode)(common_1.HttpStatus.CREATED),
    __param(0, (0, common_1.UploadedFiles)()),
    __param(1, (0, common_1.Req)()),
    __param(2, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object, Object]),
    __metadata("design:returntype", Promise)
], ServiceHistoryController.prototype, "create", null);
__decorate([
    (0, common_1.UseGuards)(auth_guard_1.AuthGuard),
    (0, common_1.Get)(),
    (0, common_1.HttpCode)(common_1.HttpStatus.OK),
    __param(0, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], ServiceHistoryController.prototype, "all", null);
__decorate([
    (0, common_1.UseGuards)(auth_guard_1.AuthGuard),
    (0, common_1.Get)(':id'),
    (0, common_1.HttpCode)(common_1.HttpStatus.OK),
    __param(0, (0, common_1.Req)()),
    __param(1, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], ServiceHistoryController.prototype, "detail", null);
__decorate([
    (0, common_1.UseGuards)(auth_guard_1.AuthGuard),
    (0, common_1.Delete)(':id'),
    (0, common_1.HttpCode)(common_1.HttpStatus.OK),
    __param(0, (0, common_1.Req)()),
    __param(1, (0, common_1.Res)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], ServiceHistoryController.prototype, "destroy", null);
exports.ServiceHistoryController = ServiceHistoryController = ServiceHistoryController_1 = __decorate([
    (0, common_1.Controller)('services/history'),
    __metadata("design:paramtypes", [history_repository_1.HistoryRepository,
        history_service_1.HistoryService])
], ServiceHistoryController);
//# sourceMappingURL=history-controller.js.map