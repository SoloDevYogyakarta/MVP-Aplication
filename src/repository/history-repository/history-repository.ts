import { Injectable, Logger } from '@nestjs/common';
import { sequelize } from '../../database/entities/entity';
import {
  userAssociate,
  userAttribute,
  userHistoryInclude,
} from '../../database/associates/user-associate/user-associate';
import { UserInstance } from '../../database/entities/authenticate/user-entity/user-entity';
import { OrderEntity } from '../../database/entities/services/order-entity/order-entity';

@Injectable()
export class HistoryRepository {
  private readonly logger = new Logger(HistoryRepository.name);

  async findAll(): Promise<UserInstance[]> {
    this.logger.log(HistoryRepository.name);
    let result = await userAssociate.findAll({
      attributes: userAttribute,
      include: userHistoryInclude,
    });
    for (const values of result) {
      const visit = await this.visit(values.id);
      const total = await this.total(values.id);
      const totalOrder = await this.totalOrder(values.id);
      result = JSON.parse(JSON.stringify(result)).map((item: UserInstance) => {
        if (item.id === values.id) {
          item['visit'] = Number(visit);
          item['total'] = total;
        }
        if (totalOrder.length) {
          item['order'] = (
            item as unknown as { order: OrderEntity[] }
          ).order.map((ord) => {
            const total_order = totalOrder.find((i) => i.id === ord.id);
            if (total_order) {
              if (item.id === ord.user_id) {
                ord['total'] = total_order?.total
                  ? Number(total_order.total)
                  : 0;
              }
            }
            return ord;
          });
        }
        return item;
      });
    }
    return result;
  }

  async findOne(id: number): Promise<UserInstance> {
    this.logger.log(HistoryRepository.name);
    let result = await userAssociate.findOne({
      where: { id },
      attributes: userAttribute,
      include: userHistoryInclude,
    });
    result = JSON.parse(JSON.stringify(result));
    for (const values of (result as unknown as { order: OrderEntity[] })
      .order) {
      const total = await this.totalOrder(values.id);
      result['order'] = (
        result as unknown as { order: OrderEntity[] }
      ).order.map((item) => {
        if (item.id === values.id) {
          item['total'] = total;
        }
        return item;
      });
    }
    result['visit'] = Number(await this.visit(id));
    result['total'] = await this.total(id);
    return result;
  }

  async visit(id: number): Promise<number> {
    const query = await sequelize.query(
      `SELECT a.user_id, COUNT(a.user_id) as visit FROM "PRODUCTS"."ORDER" AS a WHERE a.user_id = '${id}' GROUP BY a.user_id `,
    );
    const visit = query[0].find(
      (item: { user_id: number; visit: number }) =>
        Number(item.user_id) === Number(id),
    ) as { user_id: number; visit: number };
    return visit?.visit ? visit.visit : 0;
  }

  async totalOrder(
    id: number,
  ): Promise<{ id: number; user_id: number; total: string }[]> {
    const query = await sequelize.query(`
    SELECT a.id,a.user_id,SUM(CAST(b.price as int)) as total FROM "PRODUCTS"."ORDER" AS a
    LEFT JOIN "SERVICES"."HISTORY" AS b ON b.order_id = a.id WHERE a.user_id = ${id}
    GROUP BY a.id;
    `);
    return query[0] as unknown as {
      id: number;
      user_id: number;
      total: string;
    }[];
  }

  async total(id: number): Promise<number> {
    const query = await sequelize.query(`
      SELECT a.user_id,SUM(CAST(b.price as int)) as total FROM "PRODUCTS"."ORDER" AS a
      LEFT JOIN "SERVICES"."HISTORY" AS b ON b.order_id = a.id WHERE a.user_id = ${id}
      GROUP BY a.id
    `);
    const result = query[0]
      .map((item: { total: string }) => (item.total === null ? 0 : item.total))
      ?.reduce((arr: unknown, curr: unknown) => {
        return Number(arr) + Number(curr);
      }, 0);
    return result ? Number(result) : 0;
  }
}
