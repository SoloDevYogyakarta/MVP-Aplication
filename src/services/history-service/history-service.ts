import { HttpException, HttpStatus, Injectable, Logger } from '@nestjs/common';
import { userEntity } from '../../database/entities/authenticate/user-entity/user-entity';
import { historyEntity } from '../../database/entities/services/history-entity/history-entity';
import { CreateHistoryField } from '../../dto/history-dto/history-dto';

@Injectable()
export class HistoryService {
  private readonly logger = new Logger(HistoryService.name);

  async create(public_id: string, body: CreateHistoryField[]) {
    this.logger.log(HistoryService.name);
    const findOne = await userEntity.findOne({ where: { public_id } });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Account not found',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    for (const values of body) {
      const create = await historyEntity.create({
        ...values,
        user_id: findOne.public_id,
      });
      create.save();
    }
    return { status: HttpStatus.CREATED, message: 'History has been added' };
  }

  async destroy(public_id: string) {
    const findOne = await historyEntity.findOne({ where: { public_id } });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'History not found',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    findOne.destroy();
    return { status: HttpStatus.OK, message: 'History has been delete' };
  }
}
