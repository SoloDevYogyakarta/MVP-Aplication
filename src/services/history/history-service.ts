import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { userEntity } from '../../database/entities/authenticates/user-entity/user-entity';
import { historyEntity } from '../../database/entities/services/history-entity/history-entity';

@Injectable()
export class HistoryService {
  async destroy(public_id: string, user_id: string) {
    await this.isAdmin(user_id);
    const findOne = await historyEntity.findOne({ where: { public_id } });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.NOT_FOUND,
          message: 'Not Found',
        },
        HttpStatus.NOT_FOUND,
      );
    }
    findOne.destroy();
    return { status: HttpStatus.OK, message: 'History has been delete' };
  }

  async isAdmin(public_id: string): Promise<void> {
    const findOne = await userEntity.findOne({ where: { public_id } });
    if (findOne?.role !== 'admin') {
      throw new HttpException('false', HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }
}
