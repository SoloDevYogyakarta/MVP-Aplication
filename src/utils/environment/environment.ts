import { config } from 'dotenv';

config();

export const environment: NodeJS.ProcessEnv = process.env;
