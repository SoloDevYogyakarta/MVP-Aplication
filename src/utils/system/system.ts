import fs from 'fs';
import { join } from 'path';

function joinpath(path: string): string {
  return join(__dirname, path);
}

function readpath(path: string): string {
  return fs.readFileSync(joinpath(path), { encoding: 'utf-8' });
}

function createpath(path: string, data: unknown): void {
  let values = data;
  if (typeof values !== 'string') {
    values = JSON.stringify(values);
  }
  fs.writeFileSync(joinpath(path), values as string);
}

function removepath(path: string): void {
  fs.unlinkSync(joinpath(path));
}

export { joinpath, readpath, createpath, removepath };
