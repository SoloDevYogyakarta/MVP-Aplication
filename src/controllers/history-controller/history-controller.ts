import {
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Logger,
  Post,
  Req,
  Res,
  UploadedFiles,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';
import { Response } from 'express';
import { omit } from 'lodash';
import { HistoryService } from '../../services/history-service/history-service';
import { HistoryRepository } from '../../repository/history-repository/history-repository';
import { AuthGuard } from '../../middleware/auth-guard/auth-guard';
import { CustomRequest } from '../../types/custom-request.type';
import { FilesInterceptor } from '@nestjs/platform-express';
import { uploadOptions } from '../../utils/upload/upload';

@Controller('history')
export class ServiceHistoryController {
  private readonly logger = new Logger(ServiceHistoryController.name);

  constructor(
    private readonly repository: HistoryRepository,
    private readonly service: HistoryService,
  ) {}

  @UseGuards(AuthGuard)
  @Post(':id')
  @UseInterceptors(FilesInterceptor('file', 5, { storage: uploadOptions }))
  @HttpCode(HttpStatus.CREATED)
  async create(
    @UploadedFiles() files: Express.Multer.File[],
    @Req() req: CustomRequest,
    @Res() res: Response,
  ) {
    this.logger.log(ServiceHistoryController.name);
    const result = await this.service.create(
      req.params.id,
      req.body.desc,
      JSON.parse(req.body.data),
      files,
    );
    return res.status(result.status).json(omit(result, ['result']));
  }

  @UseGuards(AuthGuard)
  @Get()
  @HttpCode(HttpStatus.OK)
  async all(@Res() res: Response) {
    this.logger.log(ServiceHistoryController.name);
    const result = await this.repository.findAll();
    return res.status(HttpStatus.OK).json(result);
  }

  @UseGuards(AuthGuard)
  @Get(':id')
  @HttpCode(HttpStatus.OK)
  async detail(@Req() req: CustomRequest, @Res() res: Response) {
    this.logger.log(ServiceHistoryController.name);
    const result = await this.repository.findOne(req.params.id);
    return res.status(HttpStatus.OK).json(result);
  }

  @UseGuards(AuthGuard)
  @Delete(':id')
  @HttpCode(HttpStatus.OK)
  async destroy(@Req() req: CustomRequest, @Res() res: Response) {
    const result = await this.service.destroy(req.params.id);
    return res.status(result.status).json(result);
  }
}
