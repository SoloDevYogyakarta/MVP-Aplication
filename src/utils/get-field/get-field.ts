import { GetFieldType } from 'src/types/get-field.type';
import { readpath } from '../system/system';

export function getField(filename?: GetFieldType) {
  try {
    if (filename === 'fake') {
      return readpath(`../../database/dataTxt/${filename}.txt`);
    }
    return JSON.parse(readpath(`../../database/dataTxt/${filename}.txt`));
  } catch (err) {
    throw new Error(err);
  }
}
