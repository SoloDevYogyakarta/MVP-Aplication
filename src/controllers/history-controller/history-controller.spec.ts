import { HttpStatus, INestApplication } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { Test } from '@nestjs/testing';
import { getfield } from '../../utils/get-field/get-field';
import { HistoryRepository } from '../../repository/history-repository/history-repository';
import { HistoryService } from '../../services/history-service/history-service';
import { environment } from '../../utils/environment/environment';
import { ServiceHistoryController } from './history-controller';
import supertest from 'supertest';
import { faker } from '@faker-js/faker';
import { createpath, joinpath } from '../../utils/system/system';

describe('ServiceHistoryController', () => {
  let destory_id!: number;
  let user_id!: number;
  let token!: string;
  let app: INestApplication;

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        JwtModule.register({
          secret: environment['SECRET'],
          signOptions: {
            expiresIn: '7d',
          },
        }),
      ],
      controllers: [ServiceHistoryController],
      providers: [HistoryService, HistoryRepository],
    }).compile();

    app = moduleRef.createNestApplication();
    await app.init();
  });

  it('should to be defined', () => expect(app).toBeDefined());

  it('render correctly', () => expect(app.getHttpServer()).toMatchSnapshot());

  try {
    token = getfield('token');
  } catch (err) {
    // empty
  }
  try {
    user_id = getfield('user-http-entity').id;
  } catch (err) {
    // empty
  }
  try {
    destory_id = getfield('history-entity').id;
  } catch (err) {
    // empty
  }

  if (user_id && token) {
    it('http::history create', async () => {
      await supertest(app.getHttpServer())
        .post(`/services/history/${user_id}`)
        .set('Authorization', `Bearer ${token}`)
        .field('desc', faker.lorem.paragraph())
        .field(
          'data',
          JSON.stringify([
            {
              name: faker.commerce.productName(),
              title: faker.commerce.productMaterial(),
              desc: faker.commerce.productDescription(),
              price: faker.number.int({ min: 1000, max: 20000 }),
            },
          ]),
        )
        .attach(
          'file',
          joinpath(
            '../../../391282393_7054748857952078_2554999196306250130_n.jpg',
          ),
        )
        .attach(
          'file',
          joinpath(
            '../../../391282393_7054748857952078_2554999196306250130_n.jpg',
          ),
        )
        .expect(HttpStatus.CREATED)
        .then((res) => {
          expect(res.body).toEqual({
            status: HttpStatus.CREATED,
            message: 'History has been added',
          });
        });
    });

    it('http::history invalid create', async () => {
      await supertest(app.getHttpServer())
        .post(`/services/history/dqwdqwdq`)
        .set('Authorization', `Bearer ${token}`)
        .field(
          'data',
          JSON.stringify([
            {
              name: faker.commerce.productName(),
              title: faker.commerce.productMaterial(),
              desc: faker.commerce.productDescription(),
              price: faker.number.int({ min: 1000, max: 20000 }),
            },
          ]),
        )
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.BAD_REQUEST,
            message: 'Account not found',
          }),
        );
    });

    it('http::history all', async () => {
      await supertest(app.getHttpServer())
        .get('/services/history')
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.OK)
        .then((res) => {
          expect(res.body.length).not.toEqual(0);
        });
    });

    it('http::history detail', async () => {
      await supertest(app.getHttpServer())
        .get(`/services/history/${user_id}`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .then((res) => {
          createpath('../folder-text/history-http-entity.txt', res.body);
          expect(res.body.id).toEqual(user_id);
        });
    });
  }

  if (destory_id) {
    it('http::history destroy', async () => {
      await supertest(app.getHttpServer())
        .delete(`/services/history/${destory_id}`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.OK)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.OK,
            message: 'History has been delete',
          }),
        );
    });

    it('http::history invalid destroy', async () => {
      await supertest(app.getHttpServer())
        .delete(`/services/history/dqwdqwdq`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.BAD_REQUEST,
            message: 'History not found',
          }),
        );
    });
  }
});
