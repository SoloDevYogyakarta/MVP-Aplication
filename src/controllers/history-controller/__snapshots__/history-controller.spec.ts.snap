// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceHistoryController render correctly 1`] = `
Server {
  "_connections": 0,
  "_events": {
    "connection": [Function],
    "request": [Function],
  },
  "_eventsCount": 2,
  "_handle": null,
  "_maxListeners": undefined,
  "_unref": false,
  "_usingWorkers": false,
  "_workers": [],
  "allowHalfOpen": true,
  "connectionsCheckingInterval": 30000,
  "headersTimeout": 60000,
  "highWaterMark": 16384,
  "httpAllowHalfOpen": false,
  "insecureHTTPParser": undefined,
  "joinDuplicateHeaders": undefined,
  "keepAlive": false,
  "keepAliveInitialDelay": 0,
  "keepAliveTimeout": 5000,
  "maxHeaderSize": undefined,
  "maxHeadersCount": null,
  "maxRequestsPerSocket": 0,
  "noDelay": true,
  "pauseOnConnect": false,
  "rejectNonStandardBodyWrites": false,
  "requestTimeout": 300000,
  "timeout": 0,
  Symbol(IncomingMessage): [Function],
  Symbol(ServerResponse): [Function],
  Symbol(kCapture): false,
  Symbol(async_id_symbol): -1,
  Symbol(http.server.connections): ConnectionsList {},
  Symbol(http.server.connectionsCheckingInterval): Timeout {
    "_destroyed": false,
    "_idleNext": Timeout {
      "_destroyed": false,
      "_idleNext": Timeout {
        "_destroyed": false,
        "_idleNext": Timeout {
          "_destroyed": false,
          "_idleNext": TimersList {
            "_idleNext": [Circular],
            "_idlePrev": [Circular],
            "expiry": 34599,
            "id": -9007199254740988,
            "msecs": 30000,
            "priorityQueuePosition": 2,
          },
          "_idlePrev": [Circular],
          "_idleStart": 4599,
          "_idleTimeout": 30000,
          "_onTimeout": [Function],
          "_repeat": 30000,
          "_timerArgs": undefined,
          Symbol(refed): false,
          Symbol(kHasPrimitive): false,
          Symbol(asyncId): 19,
          Symbol(triggerId): 0,
        },
        "_idlePrev": [Circular],
        "_idleStart": 4615,
        "_idleTimeout": 30000,
        "_onTimeout": [Function],
        "_repeat": 30000,
        "_timerArgs": undefined,
        Symbol(refed): false,
        Symbol(kHasPrimitive): false,
        Symbol(asyncId): 23,
        Symbol(triggerId): 0,
      },
      "_idlePrev": [Circular],
      "_idleStart": 6704,
      "_idleTimeout": 30000,
      "_onTimeout": [Function],
      "_repeat": 30000,
      "_timerArgs": undefined,
      Symbol(refed): false,
      Symbol(kHasPrimitive): false,
      Symbol(asyncId): 390,
      Symbol(triggerId): 0,
    },
    "_idlePrev": TimersList {
      "_idleNext": [Circular],
      "_idlePrev": Timeout {
        "_destroyed": false,
        "_idleNext": [Circular],
        "_idlePrev": Timeout {
          "_destroyed": false,
          "_idleNext": [Circular],
          "_idlePrev": Timeout {
            "_destroyed": false,
            "_idleNext": [Circular],
            "_idlePrev": [Circular],
            "_idleStart": 6704,
            "_idleTimeout": 30000,
            "_onTimeout": [Function],
            "_repeat": 30000,
            "_timerArgs": undefined,
            Symbol(refed): false,
            Symbol(kHasPrimitive): false,
            Symbol(asyncId): 390,
            Symbol(triggerId): 0,
          },
          "_idleStart": 4615,
          "_idleTimeout": 30000,
          "_onTimeout": [Function],
          "_repeat": 30000,
          "_timerArgs": undefined,
          Symbol(refed): false,
          Symbol(kHasPrimitive): false,
          Symbol(asyncId): 23,
          Symbol(triggerId): 0,
        },
        "_idleStart": 4599,
        "_idleTimeout": 30000,
        "_onTimeout": [Function],
        "_repeat": 30000,
        "_timerArgs": undefined,
        Symbol(refed): false,
        Symbol(kHasPrimitive): false,
        Symbol(asyncId): 19,
        Symbol(triggerId): 0,
      },
      "expiry": 34599,
      "id": -9007199254740988,
      "msecs": 30000,
      "priorityQueuePosition": 2,
    },
    "_idleStart": 6713,
    "_idleTimeout": 30000,
    "_onTimeout": [Function],
    "_repeat": 30000,
    "_timerArgs": undefined,
    Symbol(refed): false,
    Symbol(kHasPrimitive): false,
    Symbol(asyncId): 394,
    Symbol(triggerId): 0,
  },
  Symbol(kUniqueHeaders): null,
}
`;
